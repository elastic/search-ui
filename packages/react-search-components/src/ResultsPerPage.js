import PropTypes from "prop-types";
import React from "react";
import Select from "react-select";

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({})
};

function ResultsPerPage({ onChange, options, value }) {
  const selectedValue = value;

  const selectedOption = selectedValue
    ? options.find(option => option === selectedValue)
    : null;

  return (
    <div className="sui-results-per-page">
      <div className="sui-results-per-page__label">Show</div>
      <Select
        className="sui-select sui-select--inline"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={o => onChange(o)}
        options={options}
        isSearchable={false}
        styles={setDefaultStyle}
      />
    </div>
  );
}

ResultsPerPage.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.number).isRequired,
  value: PropTypes.number
};

export default ResultsPerPage;
