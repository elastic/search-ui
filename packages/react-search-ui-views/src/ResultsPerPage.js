import PropTypes from "prop-types";
import React from "react";
import Select from "react-select";

import { appendClassName } from "./view-helpers";

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({}),
  singleValue: provided => {
    // Pulling out CSS that we don't want
    // eslint-disable-next-line no-unused-vars
    const { position, top, transform, maxWidth, ...rest } = provided;
    return { ...rest, lineHeight: 1, marginRight: 0 };
  },
  valueContainer: provided => ({ ...provided, paddingRight: 0 })
};

const wrapOption = option => ({ label: option, value: option });

function ResultsPerPage({
  className,
  onChange,
  options,
  value: selectedValue,
  ...rest
}) {
  let selectedOption = null;

  if (selectedValue) {
    selectedOption = wrapOption(selectedValue);

    if (!options.includes(selectedValue)) options = [selectedValue, ...options];
  }

  return (
    <div
      className={appendClassName("sui-results-per-page", className)}
      {...rest}
    >
      <div className="sui-results-per-page__label">Show</div>
      <Select
        className="sui-select sui-select--inline"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={o => onChange(o.value)}
        options={options.map(wrapOption)}
        isSearchable={false}
        styles={setDefaultStyle}
      />
    </div>
  );
}

ResultsPerPage.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.number).isRequired,
  className: PropTypes.string,
  value: PropTypes.number
};

export default ResultsPerPage;
