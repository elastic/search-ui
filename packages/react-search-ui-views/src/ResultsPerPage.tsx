import type { SearchContextState } from "@elastic/search-ui";
import React from "react";
import Select, { components } from "react-select";
import { BaseContainerProps, Rename } from "./types";

import { appendClassName } from "./view-helpers";

export type ResultsPerPageContainerContext = Pick<
  SearchContextState,
  "resultsPerPage" | "setResultsPerPage"
>;

export type ResultsPerPageContainerProps = BaseContainerProps &
  ResultsPerPageContainerContext & {
    view?: React.ComponentType<ResultsPerPageViewProps>;
    options?: number[];
  };

export type ResultsPerPageViewProps = BaseContainerProps &
  Pick<ResultsPerPageContainerProps, "options"> &
  Rename<
    ResultsPerPageContainerContext,
    {
      setResultsPerPage: "onChange";
      resultsPerPage: "value";
    }
  >;

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({}),
  singleValue: (provided) => {
    // Pulling out CSS that we don't want
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { position, top, transform, maxWidth, ...rest } = provided;
    return { ...rest, lineHeight: 1, marginRight: 0 };
  },
  valueContainer: (provided) => ({ ...provided, paddingRight: 0 })
};

const wrapOption = (option) => ({ label: option, value: option });

function Option(props) {
  return <components.Option {...props}>{props.data.label}</components.Option>;
}

function ResultsPerPage({
  className,
  onChange,
  options,
  value: selectedValue,
  ...rest
}: ResultsPerPageViewProps) {
  let selectedOption = null;

  if (selectedValue) {
    selectedOption = wrapOption(selectedValue);

    if (!options.includes(selectedValue)) options = [selectedValue, ...options];
  }

  return (
    <div
      className={appendClassName("sui-results-per-page", className)}
      {...rest}
    >
      <div className="sui-results-per-page__label">Show</div>
      <Select
        className="sui-select sui-select--inline"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={(o) => onChange(o.value)}
        options={options.map(wrapOption)}
        isSearchable={false}
        styles={setDefaultStyle}
        components={{
          Option: (props) => {
            props.innerProps["data-transaction-name"] = `results per page`;
            return Option(props);
          }
        }}
      />
    </div>
  );
}

export default ResultsPerPage;
