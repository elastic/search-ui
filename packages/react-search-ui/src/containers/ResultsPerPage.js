import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { ResultsPerPage } from "@elastic/react-search-ui-views";

export class ResultsPerPageContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    view: PropTypes.func,
    options: PropTypes.arrayOf(PropTypes.number),
    // State
    resultsPerPage: PropTypes.number.isRequired,
    // Actions
    setResultsPerPage: PropTypes.func.isRequired
  };

  static defaultProps = {
    options: [20, 40, 60]
  };

  render() {
    const {
      className,
      resultsPerPage,
      setResultsPerPage,
      view,
      options,
      ...rest
    } = this.props;

    const View = view || ResultsPerPage;

    return View({
      className,
      onChange: value => {
        setResultsPerPage(value);
      },
      options,
      value: resultsPerPage,
      ...rest
    });
  }
}

export default withSearch(({ resultsPerPage, setResultsPerPage }) => ({
  resultsPerPage,
  setResultsPerPage
}))(ResultsPerPageContainer);
