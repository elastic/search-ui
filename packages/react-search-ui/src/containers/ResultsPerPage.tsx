import React, { Component } from "react";
import { withSearch } from "..";
import {
  ResultsPerPage,
  ResultsPerPageContainerContext,
  ResultsPerPageContainerProps,
  ResultsPerPageViewProps
} from "@elastic/react-search-ui-views";

export class ResultsPerPageContainer extends Component<ResultsPerPageContainerProps> {
  static defaultProps = {
    options: [20, 40, 60]
  };

  render() {
    const {
      className,
      resultsPerPage,
      setResultsPerPage,
      options,
      view,
      ...rest
    } = this.props;

    const View = view || ResultsPerPage;
    const viewProps: ResultsPerPageViewProps = {
      className,
      onChange: (value) => {
        setResultsPerPage(value);
      },
      options,
      value: resultsPerPage,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<
  ResultsPerPageContainerProps,
  ResultsPerPageContainerContext
>(({ resultsPerPage, setResultsPerPage }) => ({
  resultsPerPage,
  setResultsPerPage
}))(ResultsPerPageContainer);
