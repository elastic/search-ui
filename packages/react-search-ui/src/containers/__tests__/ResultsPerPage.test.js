import React from "react";
import { shallow } from "enzyme";
import { ResultsPerPageContainer } from "../ResultsPerPage";

const params = {
  results: [{}, {}],
  resultsPerPage: 20,
  searchTerm: "test",
  setResultsPerPage: jest.fn()
};

beforeEach(() => {
  params.setResultsPerPage = jest.fn();
});

it("supports a render prop", () => {
  // eslint-disable-next-line react/prop-types
  const render = ({ value }) => {
    return <div>{value}</div>;
  };
  const wrapper = shallow(
    <ResultsPerPageContainer {...params} view={render} />
  );
  expect(wrapper).toMatchSnapshot();
});

it("renders when it doesn't have any results or a search term", () => {
  const wrapper = shallow(
    <ResultsPerPageContainer
      {...{
        ...params,
        searchTerm: "",
        results: []
      }}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("will call back when a selection is made in the view", () => {
  let viewProps;

  shallow(
    <ResultsPerPageContainer {...params} view={props => (viewProps = props)} />
  );

  const { onChange } = viewProps;
  onChange(40);

  const resultsPerPage = params.setResultsPerPage.mock.calls[0][0];
  expect(resultsPerPage).toEqual(40);
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <ResultsPerPageContainer
      {...params}
      className={className}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("renders the component with custom page options", () => {
  const options = [5, 10, 15];
  const resultsPerPage = 10;
  const wrapper = shallow(
    <ResultsPerPageContainer {...{ ...params, resultsPerPage, options }} />
  );

  expect(wrapper).toMatchSnapshot();
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <ResultsPerPageContainer
      {...params}
      data-foo={data}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});
