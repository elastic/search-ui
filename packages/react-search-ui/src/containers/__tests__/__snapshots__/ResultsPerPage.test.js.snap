// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component with custom page options 1`] = `
<div
  className="sui-results-per-page"
  results={
    Array [
      Object {},
      Object {},
    ]
  }
  searchTerm="test"
>
  <div
    className="sui-results-per-page__label"
  >
    Show
  </div>
  <StateManager
    className="sui-select sui-select--inline"
    classNamePrefix="sui-select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": 5,
          "value": 5,
        },
        Object {
          "label": 10,
          "value": 10,
        },
        Object {
          "label": 15,
          "value": 15,
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
        "singleValue": [Function],
        "valueContainer": [Function],
      }
    }
    value={
      Object {
        "label": 10,
        "value": 10,
      }
    }
  />
</div>
`;

exports[`renders when it doesn't have any results or a search term 1`] = `
<div
  className="sui-results-per-page"
  results={Array []}
  searchTerm=""
>
  <div
    className="sui-results-per-page__label"
  >
    Show
  </div>
  <StateManager
    className="sui-select sui-select--inline"
    classNamePrefix="sui-select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": 20,
          "value": 20,
        },
        Object {
          "label": 40,
          "value": 40,
        },
        Object {
          "label": 60,
          "value": 60,
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
        "singleValue": [Function],
        "valueContainer": [Function],
      }
    }
    value={
      Object {
        "label": 20,
        "value": 20,
      }
    }
  />
</div>
`;

exports[`supports a render prop 1`] = `
<div>
  20
</div>
`;
